### PR TITLE
Add VEP metadata json + service to use in java projects

### PR DESCRIPTION
**Pull request by bartcharbon**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -11,9 +11,9 @@
     <relativePath/> <!-- lookup parent from repository -->
   </parent>
 
-  <groupId>org.molgenis</groupId>
+  <groupId>com.github.molgenis</groupId>
   <artifactId>vip-utils</artifactId>
-  <version>1.0.0</version>
+  <version>v1.1.0</version>
```
Comment (dennishendriksen) [inline pom.xml:16]: 1.1.0 instead of v1.1.0

File: src/main/java/org/molgenis/vcf/utils/PedReader.java [hunk 2]
```diff
@@ -35,6 +35,7 @@
  *   <li>Genotypes
  * </ul>
  */
+@SuppressWarnings("java:S6201")
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/PedReader.java]: warning can be resolved by:
```java
this.bufferedReader = reader instanceof BufferedReader aBufferedReader ? aBufferedReader : new BufferedReader(reader);
```
see: https://rules.sonarsource.com/java/tag/java16/RSPEC-6201

File: src/main/resources/vepMetadata.json [hunk 3]
```diff
@@ -0,0 +1,351 @@
+{
+  "vepMetadata": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: Custom metadata is limited to VEP here. How about making it more generic so we can support use cases such as SnpEff, Annovar or annotations pulled up from CSQ to the INFO level (e.g. we should do this for gnomAD).

The metadata could then look like:
```json
{"CSQ": {},"gnomAD": {}}
```
with a file rename to `vepMetadata.json` to `metadata.json`.

File: src/main/resources/vepMetadata.json [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    "Consequence": {
+      "label": "Effect",
+      "description": "Effect(s) described as Sequence Ontology term(s)",
+      "numberType": "OTHER",
+      "separator": "&",
+      "type": "STRING"
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: Should/can this be CATEGORICAL?

File: src/main/resources/vepMetadata.json [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+    },
+    "InheritanceModesGene": {
+      "label": "Inh.Pat.",
+      "description": "Inheritance pattern",
+      "numberType": "OTHER",
+      "separator": "&",
+      "type": "STRING"
+    },
+    "CLIN_SIG": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: Duplicate with `clinVar_CLNSIG`?

File: src/main/resources/vepMetadata.json [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+      "type": "CATEGORICAL",
+      "categories": [
+        "B",
+        "LB",
+        "VUS",
+        "LP",
+        "P"
+      ]
+    },
+    "UMCG_CL": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: Is the latest VIP still producing this field? Can be removed?

File: src/main/resources/vepMetadata.json [hunk 7]
```diff
@@ -0,0 +1,351 @@
+{
+  "vepMetadata": {
+    "Consequence": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: `required=true`?

File: src/main/resources/vepMetadata.json [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+      "label": "Effect",
+      "description": "Effect(s) described as Sequence Ontology term(s)",
+      "numberType": "OTHER",
+      "separator": "&",
+      "type": "STRING"
+    },
+    "VIPP": {
+      "label": "VIP path",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: `required=true`?

File: src/main/resources/vepMetadata.json [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+      "categories": [
+        "B",
+        "LB",
+        "VUS",
+        "LP",
+        "P"
+      ]
+    },
+    "Feature_type": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+      "type": "CATEGORICAL",
+      "categories": [
+        "Transcript",
+        "RegulatoryFeature",
+        "MotifFeature"
+      ],
+      "required": true
+    },
+    "IMPACT": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+    },
+    "PUBMED": {
+      "label": "PubMed",
+      "description": "PubMed citations",
+      "numberType": "OTHER",
+      "separator": "&",
+      "type": "INTEGER"
+    },
+    "HGNC_ID": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "cDNA_position": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "CDS_position": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "Protein_position": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "PICK": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "ALLELE_NUM": {
+      "numberType": "NUMBER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "Existing_variation": {
+      "numberType": "OTHER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+      "separator": "&",
+      "type": "STRING"
+    },
+    "FLAGS": {
+      "numberType": "OTHER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+      "separator": "&",
+      "type": "STRING"
+    },
+    "HPO": {
+      "numberType": "OTHER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+      "separator": "&",
+      "type": "STRING"
+    },
+    "SOMATIC": {
+      "numberType": "OTHER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+      "separator": "&",
+      "type": "INTEGER"
+    },
+    "PHENO": {
+      "numberType": "OTHER",
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+      "separator": "&",
+      "type": "INTEGER"
+    },
+    "gnomAD_AFR_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_AMR_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_ASJ_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_EAS_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_FIN_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_NFE_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 28]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_OTH_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 29]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "gnomAD_SAS_AF": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: not used in the latest VIP version: remove?

File: src/main/resources/vepMetadata.json [hunk 30]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "SpliceAI_pred_DS_AG": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 31]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "SpliceAI_pred_DS_AL": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "SpliceAI_pred_DS_DG": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 33]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "SpliceAI_pred_DS_DL": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 34]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "PolyPhen": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json [hunk 35]
```diff
@@ -0,0 +1,351 @@
...
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "FLOAT"
+    },
+    "SIFT": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/resources/vepMetadata.json (hunk 3 above)
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: how about alphabetic sorting of keys within this object to make it easier to locate stuff?

File: src/main/resources/vepMetadata.json [hunk 36]
```diff
@@ -0,0 +1,351 @@
...
+      "type": "STRING"
+    },
+    "gnomAD_HN": {
+      "label": "gnomAD HN",
+      "description": "gnomAD number of homozygotes",
+      "numberType": "NUMBER",
+      "numberCount": 1,
+      "type": "INTEGER"
+    },
+    "STRAND": {
```
Comment (dennishendriksen) [inline src/main/resources/vepMetadata.json]: missing `label` and `description`.

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 37]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import lombok.experimental.SuperBuilder;
+
+@Data
+@SuperBuilder
+@NoArgsConstructor
+@EqualsAndHashCode
+public class FieldImpl implements Field {
+  @NonNull
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: Unrelated: I noticed that you can use `@NonNull` on constructor arguments: https://projectlombok.org/features/NonNull. This could replace all the `requireNonNull` checks in our code base.

File: src/main/java/org/molgenis/vcf/utils/model/Field.java [hunk 38]
```diff
@@ -0,0 +1,27 @@
+package org.molgenis.vcf.utils.model;
+
+import java.util.Set;
+import org.molgenis.vcf.utils.model.FieldImpl.NumberType;
+
+public interface Field {
+
+  String getId();
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/Field.java]: add `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/Field.java [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+
+  FieldType getFieldType();
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/Field.java]: add `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/Field.java [hunk 40]
```diff
@@ -0,0 +1,27 @@
...
+
+  ValueType getType();
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/Field.java]: add `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/Field.java [hunk 41]
```diff
@@ -0,0 +1,27 @@
...
+
+  NumberType getNumberType();
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/Field.java]: add `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 42]
```diff
@@ -0,0 +1,35 @@
...
+  public enum NumberType {
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: part of the field API --> move to separate class

File: src/main/java/org/molgenis/vcf/utils/model/Field.java [hunk 43]
```diff
@@ -0,0 +1,27 @@
...
+
+  Integer getNumberCount();
+
+  Character getSeparator();
+
+  Set<String> getCategories();
+
+  boolean isRequired();
+
+  String getLabel();
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/Field.java]: Would be convenient if this is `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/Field.java [hunk 44]
```diff
@@ -0,0 +1,27 @@
...
+
+  String getDescription();
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/Field.java]: Would be convenient if this is `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 45]
```diff
@@ -0,0 +1,35 @@
...
+
+  @NonNull String id;
+  @NonNull FieldType fieldType;
+  @NonNull ValueType type;
+  @NonNull NumberType numberType;
+  Integer numberCount;
+  @Default boolean required = false;
+  Character separator;
+  Set<String> categories;
+  String label;
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: Would be convenient if this is `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 46]
```diff
@@ -0,0 +1,35 @@
...
+  String description;
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: Would be convenient if this is `@NonNull`

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 47]
```diff
@@ -0,0 +1,35 @@
...
+    PER_ALT_AND_REF,
+    PER_GENOTYPE,
+    OTHER
+  }
+
+  @NonNull String id;
+  @NonNull FieldType fieldType;
+  @NonNull ValueType type;
+  @NonNull NumberType numberType;
+  Integer numberCount;
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: how about `int numberCount` + throw an exception if the `NumberType` is not `NUMBER`?

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 48]
```diff
@@ -0,0 +1,35 @@
...
+  @Default boolean required = false;
+  Character separator;
+  Set<String> categories;
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: how about `@NonNull` + throw an exception if the `ValueType` is not `CATEGORICAL`?

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java [hunk 49]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.util.Set;
+import lombok.Builder.Default;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import lombok.experimental.SuperBuilder;
+
+@Data
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: Immutable `@Value` class would be better?

File: src/main/java/org/molgenis/vcf/utils/model/MissingField.java [hunk 50]
```diff
@@ -0,0 +1,88 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Value;
+import lombok.experimental.NonFinal;
+import org.molgenis.vcf.utils.model.FieldImpl.NumberType;
+
+@Value
+@Builder
+@NonFinal
```
Comment (dennishendriksen) [inline src/main/java/org/molgenis/vcf/utils/model/MissingField.java]: Class appears final instead of `@NonFinal`?

**Review by dennishendriksen — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 51]
```diff
@@ -11,9 +11,9 @@
     <relativePath/> <!-- lookup parent from repository -->
   </parent>
 
-  <groupId>org.molgenis</groupId>
+  <groupId>com.github.molgenis</groupId>
```
Comment (mswertz) [inline pom.xml:14]: why this change?

Comment (bartcharbon) [reply pom.xml:14]: this is related to the jitpack construction to use github releases as maven dependecies. We're looking into fixing this with a different construction for this behaviour (e.g. github packages)

File: pom.xml (hunk 1 above)
Comment (bartcharbon) [reply pom.xml:16]: this is related to easier use with jitpack, can probably be reverted if we move to github packages

File: src/main/resources/vepMetadata.json (hunk 5 above)
Comment (bartcharbon) [reply src/main/resources/vepMetadata.json]: no those are actually different field in our current output VCF

File: src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java (hunk 49 above)
Comment (bartcharbon) [reply src/main/java/org/molgenis/vcf/utils/model/FieldImpl.java]: Doesn't work properly in combination with extending this class (NestedField)

File: src/main/resources/vepMetadata.json (hunk 8 above)
Comment (bartcharbon) [reply src/main/resources/vepMetadata.json]: don't think so, adding the path is a VIP setting

File: src/main/resources/vepMetadata.json (hunk 7 above)
Comment (bartcharbon) [reply src/main/resources/vepMetadata.json]: Don't think so, we add empty CSQ's if the CSQ is missing in the decision tree to have place to store the VIPP and VIPC

**Review by dennishendriksen — CHANGES_REQUESTED** [review]
# Metadata format: proposed changes	
1) JSON schema
```json
{
  "info": {
    "key0": {},
	"key1": {}
  },
  "format": {
	"key1": {},
	"key2": {},
  }
}
```
2) rename resource from vepMetadata.json to field_metadata.json

# Data model: comments
1) remove FieldType? user knows what the type is when retrieving the field using example code above if not, then propose to change to: public enum FieldType { INFO, FORMAT }
2) MissingField doesn't belong in vip-utils? move to vip-decision-tree?
3) feels like the data models is quite different from the TypeScript data model?

# MetadataService: proposed changes
1) load once, use many
2) should work for format fields as well as info fields
3) should work for non-nested fields as well (e.g. to add categories to an INFO field)
3) terminology:
   FieldMetadataStore instead of MetadataService --> to make clear what kind of metadata
   FieldMetadata instead of NestedMetadata       --> a field can be nested or unnested

could look something like this:
```java
FieldMetadataStore fieldMetadataStore = FieldMetadataStoreFactory.load(Path path)

public interface FieldMetadataService {
  FieldMetadata getInfo(VCFInfoHeaderLine headerLine);
  
  FieldMetadata getFormat(VCFFormatHeaderLine headerLine);
}
```

# MetadataServiceImpl: comments
1) can we make it vep-unaware? 
2) separation of concerns: shouldn't it return a directly mapped JSON -> Java?
   e.g. vip decision tree can transform the Java classes to something more useful for its needs?

**Comment by bartcharbon** [comment]
The metadataService can be VEP unaware, but `vip decision tree can transform the Java classes to something more useful for its needs` kind of defeats the purpose I guess, a post processor should be part of this lib I think (e.g. as decorator) since the goal is to deduplicate code over all our libs. If we remove this completely every lib will duplicate this postprocessing code again.

**Review by dennishendriksen — APPROVED** [review]
